### PR TITLE
[feat][ShardedDDP] Support multiple groups

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Needed for #351, and good practice in general

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 1]
```diff
@@ -565,3 +572,84 @@ def test_gpt2():
...
     device = "cuda"
     mp.spawn(run_test_gpt2, args=(world_size, backend, device, temp_file_name), nprocs=world_size, join=True)
+
+
+def run_test_multiple_groups(rank, world_size, tempfile_name):
+    # Only work with the even ranks, to check that the global_rank indexing is properly used
+    dist.init_process_group(init_method="file://" + tempfile_name, backend="nccl", rank=rank, world_size=world_size)
+
+    sub_group_ranks = [0, 2]
+    process_group = torch.distributed.new_group(ranks=sub_group_ranks, backend="nccl")
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:582]: Should we test on gloo as well?

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 2]
```diff
@@ -565,3 +572,84 @@ def test_gpt2():
...
+
+    # Make sure that all the ranks get different training data
+    # So that the sync check in between their models is meaningful
+    torch.manual_seed(rank)
+    np.random.seed(rank)
+
+    # Standard deep learning setup
+    device = "cuda"
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:590]: Interesting that this works. I always thought this meant default "cuda:0" device.

File: tests/nn/data_parallel/test_sharded_ddp.py [hunk 3]
```diff
@@ -565,3 +572,84 @@ def test_gpt2():
...
+                            )
+
+    if rank in sub_group_ranks:
+        # Model fitting in the broadcast bucket
+        model = torch.nn.Sequential(torch.nn.Linear(input_width, hidden), torch.nn.Linear(hidden, target_width)).to(
+            device
+        )
+
+        # With SGD, Momentum is required to get a state to shard
+        optimizer = OSS(model.parameters(), lr=0.1, momentum=0.99, group=process_group, broadcast_buffer_size=2 ** 20)
```
Comment (anj-s) [inline tests/nn/data_parallel/test_sharded_ddp.py:631]: can we add a for loop here? I think this is the only thing that is different right?

**Review by anj-s — APPROVED** [review]

File: tests/nn/data_parallel/test_sharded_ddp.py (hunk 3 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp.py:631]: agreed, old test which could do with some refactoring, I'll do that

File: tests/nn/data_parallel/test_sharded_ddp.py (hunk 2 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp.py:590]: it works when you specify the actual device on top of that, `torch.cuda.set_device(rank)`

File: tests/nn/data_parallel/test_sharded_ddp.py (hunk 3 above)
Comment (blefaudeux) [reply tests/nn/data_parallel/test_sharded_ddp.py]: turns out that this setting is not used anymore actually, I removed the duplicates altogether

**Comment by anj-s** [comment]
Thank you for making the changes, looks great!

**Pull request merged**